File: examples/go-cli/internal/calculator/calculator.go
package calculator

import (
	"errors"
	"fmt"
)

var (
	ErrDivisionByZero   = errors.New("cannot divide by zero")
	ErrInvalidOperation = errors.New("invalid operation")
)

func Add(a, b float64) float64 {
	return a + b
}

func Subtract(a, b float64) float64 {
	return a - b
}

func Multiply(a, b float64) float64 {
	return a * b
}

func Divide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, ErrDivisionByZero
	}
	return a / b, nil
}

func Calculate(a, b float64, operation string) (float64, error) {
	switch operation {
	case "add":
		return Add(a, b), nil
	case "subtract":
		return Subtract(a, b), nil
	case "multiply":
		return Multiply(a, b), nil
	case "divide":
		return Divide(a, b)
	default:
		return 0, fmt.Errorf("%w: %s", ErrInvalidOperation, operation)
	}
}

File: examples/node-express/src/routes/calculator.js
/** Calculator API routes. */

import express from "express";
import { calculate, DivisionByZeroError, InvalidOperationError } from "../services/calculator.js";

const router = express.Router();

/**
 * @swagger
 * /calculate:
 *   post:
 *     summary: Perform a calculation
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             required: [a, b, operation]
 *             properties:
 *               a: { type: number }
 *               b: { type: number }
 *               operation: { type: string, enum: [add, subtract, multiply, divide] }
 *     responses:
 *       200:
 *         description: Calculation result
 *       400:
 *         description: Invalid operation or division by zero
 */
router.post("/", (req, res, next) => {
  try {
    const { a, b, operation } = req.body;

    // Validation
    if (typeof a !== "number" || typeof b !== "number" || typeof operation !== "string") {
      return res.status(400).json({ error: "Invalid request: a, b must be numbers, operation must be string" });
    }

    const result = calculate(a, b, operation);

    res.json({
      result,
      operation,
      a,
      b,
    });
  } catch (error) {
    next(error);
  }
});

export default router;